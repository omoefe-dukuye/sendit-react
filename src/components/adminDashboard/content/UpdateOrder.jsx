import React, { Component } from 'react';
import '@babel/polyfill';
import { toast } from 'react-toastify';
import Modal from '../../modal';
import UpdatedParcelLocation from './updatedParcelLocation';
import axios from '../../../utils/axiosConfig';

export class UpdateOrder extends Component {
  state = {
    parcelId: '',
    ParcelIdErrorMessage: '',
    locationErrorMessage: '',
    location: ''
  };

  onLocationChange = ({ target: { value: location } }) => {
    this.setState({
      location,
      locationErrorMessage: ''
    });
  };

  onLocationBlur = async () => {
    const { location } = this.state;
    if (!location) {
      return this.setState({ locationErrorMessage: 'cannot be empty.' });
    }
  }

  onParcelIdChange = ({ target: { value: parcelId } }) => {
    this.setState({
      parcelId,
      parcelIdErrorMessage: ''
    });
  };

  onParcelIdBlur = async () => {
    const { parcelId } = this.state;
    if (!Number(parcelId)) {
      return this.setState({ parcelIdErrorMessage: 'must be a number.' });
    }
  }


  handleSubmit = async (e) => {
    e.preventDefault();
    const { location, parcelId } = this.state;

    if (!location) {
      return this.setState({ locationErrorMessage: 'cannot be empty.' });
    }

    if (!Number(parcelId)) {
      return this.setState({ parcelIdErrorMessage: 'must be a number.' });
    }

    this.setState({ buttonDisabled: true });

    try {
      const { data } = await axios.patch(`${process.env.API_ROOT_URL}/parcels/${parcelId}/currentlocation`, { location });
      this.setState({ newParcel: data });
    } catch ({ response: { data: { error } } }) {
      toast.error(error);
    } 

    this.setState({ buttonDisabled: false });
  }

  removeNewParcel = () => {
    this.setState({ newParcel: false });
  }

  render() {
    const {
      state: {
        buttonDisabled,
        location,
        parcelId,
        locationErrorMessage,
        parcelIdErrorMessage,
        newParcel
      },
      onLocationChange,
      onLocationBlur,
      onParcelIdBlur,
      onParcelIdChange,
      handleSubmit,
      removeNewParcel
    } = this;

    const message = locationErrorMessage
      || parcelIdErrorMessage
      || false;


    const disableButton = buttonDisabled || message;


    return (
      <div className='form__page create-order'>
        <div className='form__container'>
          <form className='form create-order__form'>
            <div className='form__input-top'>
              <p>
                <span className='form__input-top__label'>Current Location</span>
                <span className='form__input-top__error'>{locationErrorMessage}</span>
              </p>
            </div>
            <input
              className={locationErrorMessage ? 'red-border' : ''}
              name='location'
              value={location}
              onChange={onLocationChange}
              onBlur={onLocationBlur}
              required
            />
            <div className='form__input-top'>
              <p>
                <span className='form__input-top__label'>Parcel ID</span>
                <span className='form__input-top__error'>{parcelIdErrorMessage}</span>
              </p>
            </div>
            <input
              className={parcelIdErrorMessage ? 'red-border' : ''}
              type='number'
              name='parcelId'
              value={parcelId}
              onChange={onParcelIdChange}
              onBlur={onParcelIdBlur}
              required
            />
            <button
              disabled={disableButton}
              className='button create-order__form__button'
              onClick={handleSubmit}
            >
              <span className={`button__text--create-order ${buttonDisabled ? 'hidden' : undefined}`}>
                Update Order
              </span>
              <span className={buttonDisabled ? 'spinner spinner--button' : undefined}></span>
            </button>
          </form>
        </div>
        {
          !!newParcel
            &&
              <Modal
              isOpen={!!newParcel}
              hideModal={removeNewParcel}
              content={<UpdatedParcelLocation parcel={newParcel} />}
              />
        }
      </div>
    );
  }
}

export default UpdateOrder;
