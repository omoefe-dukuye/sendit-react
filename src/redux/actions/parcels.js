import { toast } from "react-toastify";
import axios from "../../utils/axiosConfig";
import { ADD_ORDER } from "./types";

const api = process.env.API_ROOT_URL;

export const createOrder = (payload) => async (dispatch) => {
  try {
    const { data: { parcel } } = await axios.post(`${api}/parcels`, payload);

    dispatch({
      type: ADD_ORDER,
      payload: parcel
    });

    return parcel;
  } catch ({ response: { data: { error } } }) {
   toast.error(error);
  }
}

export default createOrder;
