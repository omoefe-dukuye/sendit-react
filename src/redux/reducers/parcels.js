import { ADD_ORDER } from '../actions/types';

const parcelReducerDefaultState = [];

export default (state = parcelReducerDefaultState, action) => {
  switch(action.type) {
    case ADD_ORDER:
      return [...state, action.payload];

    default:
      return state;
  }
};
