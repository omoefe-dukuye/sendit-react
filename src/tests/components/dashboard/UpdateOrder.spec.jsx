import React from 'react';
import { shallow } from 'enzyme';
import '@babel/polyfill';
import axios from '../../../utils/axiosConfig';
import parcels from '../../mocks/mockParcels';
import UpdateOrder from '../../../components/adminDashboard/content/UpdateOrder';

jest.mock('../../../utils/axiosConfig');

const tick = () => {
  return new Promise(resolve => {
    setTimeout(resolve, 0);
  })
};

test('Should render static component', async () => {
  const wrapper = shallow(<UpdateOrder />);
  expect(wrapper).toMatchSnapshot();
});

test('Should handle parcelId input interactions', async () => {
  const wrapper = shallow(<UpdateOrder />);
  wrapper.find('input').at(1).simulate('change', { target: { value: 3 } });
  expect(wrapper.state('parcelId')).toBe(3);

  wrapper.find('input').at(1).simulate('blur');
  expect(wrapper.state('parcelIdErrorMessage')).toBe('');

  wrapper.find('input').at(1).simulate('change', { target: { value: 'p' } });
  expect(wrapper.state('parcelId')).toBe('p');

  wrapper.find('input').at(1).simulate('blur');
  expect(wrapper.state('parcelIdErrorMessage')).toBe('must be a number.');
});

test('Should handle location input interactions', async () => {
  const wrapper = shallow(<UpdateOrder />);
  wrapper.find('input').at(0).simulate('blur');
  expect(wrapper.state('locationErrorMessage')).toBe('cannot be empty.');

  wrapper.find('input').at(0).simulate('change', { target: { value: 'lokoja' } });
  expect(wrapper.state('location')).toBe('lokoja');

  wrapper.find('input').at(0).simulate('blur');
  expect(wrapper.state('location')).toBe('lokoja');
});

test('Should handle submit interactions', async () => {
  axios.patch.mockRejectedValue({ response: { data: { error: 'error' } } });
  const wrapper = shallow(<UpdateOrder />);
  wrapper.find('button').simulate('click', { preventDefault: jest.fn() });
  expect(wrapper.state('locationErrorMessage')).toBe('cannot be empty.');

  wrapper.find('input').at(0).simulate('change', { target: { value: 'lokoja' } });
  expect(wrapper.state('location')).toBe('lokoja');
  wrapper.find('button').simulate('click', { preventDefault: jest.fn() });
  expect(wrapper.state('parcelIdErrorMessage')).toBe('must be a number.');

  wrapper.find('input').at(1).simulate('change', { target: { value: 3 } });
  wrapper.find('button').simulate('click', { preventDefault: jest.fn() });
  expect(wrapper.state('parcelId')).toBe(3);

  axios.patch.mockResolvedValue({ data: parcels[0] });
  wrapper.find('button').simulate('click', { preventDefault: jest.fn() });
  await tick();
  expect(wrapper.state('newParcel')).toEqual(parcels[0]);

});
