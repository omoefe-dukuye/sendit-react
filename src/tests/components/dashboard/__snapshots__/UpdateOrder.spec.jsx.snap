// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render static component 1`] = `
<div
  className="form__page create-order"
>
  <div
    className="form__container"
  >
    <form
      className="form create-order__form"
    >
      <div
        className="form__input-top"
      >
        <p>
          <span
            className="form__input-top__label"
          >
            Current Location
          </span>
          <span
            className="form__input-top__error"
          />
        </p>
      </div>
      <input
        className=""
        name="location"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        value=""
      />
      <div
        className="form__input-top"
      >
        <p>
          <span
            className="form__input-top__label"
          >
            Parcel ID
          </span>
          <span
            className="form__input-top__error"
          />
        </p>
      </div>
      <input
        className=""
        name="parcelId"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        type="number"
        value=""
      />
      <button
        className="button create-order__form__button"
        disabled={false}
        onClick={[Function]}
      >
        <span
          className="button__text--create-order undefined"
        >
          Update Order
        </span>
        <span />
      </button>
    </form>
  </div>
</div>
`;
